src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -295,6 +295,84 @@ internal static partial class Number
                                         "80818283848586878889"u8 +
                                         "90919293949596979899"u8;
 
+        public static unsafe string FormatDecimal32(Decimal32 value, ReadOnlySpan<char> format, NumberFormatInfo info)
+        {
+            char fmt = ParseFormatSpecifier(format, out int digits);
+
+            byte* pDigits = stackalloc byte[Decimal32NumberBufferLength];
+            NumberBuffer number = new NumberBuffer(NumberBufferKind.Decimal, pDigits, Decimal32NumberBufferLength);
+
+            Decimal32ToNumber(value, ref number);
+
+            char* stackPtr = stackalloc char[CharStackBufferSize];
+            var vlb = new ValueListBuilder<char>(new Span<char>(stackPtr, CharStackBufferSize));
+
+            if (fmt != 0)
+            {
+                NumberToString(ref vlb, ref number, fmt, digits, info);
+            }
+            else
+            {
+                NumberToStringFormat(ref vlb, ref number, format, info);
+            }
+
+            string result = vlb.AsSpan().ToString();
+            vlb.Dispose();
+            return result;
+        }
+
+        public static unsafe string FormatDecimal64(Decimal64 value, ReadOnlySpan<char> format, NumberFormatInfo info)
+        {
+            char fmt = ParseFormatSpecifier(format, out int digits);
+
+            byte* pDigits = stackalloc byte[Decimal64NumberBufferLength];
+            NumberBuffer number = new NumberBuffer(NumberBufferKind.Decimal, pDigits, Decimal64NumberBufferLength);
+
+            Decimal64ToNumber(value, ref number);
+
+            char* stackPtr = stackalloc char[CharStackBufferSize];
+            var vlb = new ValueListBuilder<char>(new Span<char>(stackPtr, CharStackBufferSize));
+
+            if (fmt != 0)
+            {
+                NumberToString(ref vlb, ref number, fmt, digits, info);
+            }
+            else
+            {
+                NumberToStringFormat(ref vlb, ref number, format, info);
+            }
+
+            string result = vlb.AsSpan().ToString();
+            vlb.Dispose();
+            return result;
+        }
+
+        public static unsafe string FormatDecimal128(Decimal128 value, ReadOnlySpan<char> format, NumberFormatInfo info)
+        {
+            char fmt = ParseFormatSpecifier(format, out int digits);
+
+            byte* pDigits = stackalloc byte[Decimal128NumberBufferLength];
```

Comment:
This is too large of a stack allocation. In general we want to try and limit stackalloc to no more than 1024 bytes and even then it needs to be in scenarios where we're close to a leaf method and know the total amount of stack space needed.

This on the other hand will potentially allocate a bit over 6 kilobytes and so needs to rent from the array pool or be a new allocation instead